tests/test_middleware_interface.py

```diff
@@ -43,27 +88,73 @@ def setUp(self):
                                 kind="BIAS")
         self.logger_name = "activator.middleware_interface"
 
+    def tearDown(self):
+        # Remove the temporary local butler created by the interface.
+        shutil.rmtree(self.interface.butler.datastore.root.path)
```

Comment:
I think we either don't want `MiddlewareInterface` to clean itself up (so we can do an export/import gather step on the results) or it will get cleaned up "for free" if it's local butler is a RAM disk on the node, or some other node-temporary storage.

Cleaning up everything after the test completes seems like a good idea to me, and prevents re-running tests from tripping over already existing local butler repos that didn't get removed.

Reply:
Please do not try to be clever with resources. It *will* have bugs, guaranteed.

If you intend for the repository to outlast `MiddlewareInterface` (though I would argue that if you transferring the results, you want the process object to persist until that is finished), then the object should not be responsible for creating the repository in the first place. Otherwise, it is very likely that the application code will forget to clean it up, or will run into consistency bugs relating to whether or not the repository was actually or previously created.

If you are assuming the larger framework will clean it up, then the repository should be created by that same framework. Otherwise, the implementation of `MiddlewareInterface` is making a hidden assumption that will break if, for example, somebody later decides that a RAM disk isn't going to cut it. There have been cases in other LSST packages where a bug fix broke other parts of a package because of a chain of highly specific assumptions that suddenly got invalidated.

Reply:
@kfindeisen and I talked about this, and we've changed it so that `MiddlewareInterface` takes a `butler` arg, so that the calling code does all directory creation/destruction. The tests now use `tempfile.TemporaryDirectory`, while `activator` uses the same `/tmp/repo-$PID` approach that the current code had. If we eventually want to go with a ramdisk or other specific path on gcloud (e.g. dedicated SSD), `activator` will manage that.